cpp/src/arrow/compute/kernels/filter.h

```diff
@@ -39,9 +39,31 @@ struct Datum;
 class ARROW_EXPORT FilterFunction {
  public:
   /// Filter an array with a scalar argument.
-  virtual Status Filter(const ArrayData& input, const Scalar& scalar,
+  virtual Status Filter(const ArrayData& array, const Scalar& scalar,
                         ArrayData* output) const = 0;
 
+  Status Filter(const ArrayData& array, const Scalar& scalar,
+                std::shared_ptr<ArrayData>&& output) {
```

Comment:
this seems strange why the r-value reference?

Reply:
Because of https://github.com/apache/arrow/pull/4398/files#diff-4cce1db37d9827bc657384ea82da0d63R42 . I'll modify the code to avoid the r-value.